mrmustard/math/lattice/strategies/vanilla.py

```diff
@@ -33,109 +33,195 @@
-    stable implementation because the errors are averaged out (or so we think).
+        # contribution from pivot's lower neighbours
+        # note the first is when j=0 which needs a -1 in the sqrt from delta_0j
+        value_at_index += A[0, 0] * SQRT[index[0] - 1] * G[pivot - strides[0]]
+        for j in range(1, len(strides)):
+            value_at_index += A[0, j] * SQRT[index[j]] * G[pivot - strides[j]]
+        G[flat_index] = value_at_index / SQRT[index[0]]
 
-    Supports ``b`` as a batched tensor, with the batch dimension on the first index,
-    in which case the output tensor will have the same batch dimension.
+    return G.reshape(shape)
+
+
+@njit
+def vanilla_stable(shape: tuple[int, ...], A, b, c) -> ComplexTensor:  # pragma: no cover
+    r"""Stable version of the vanilla algorithm for calculating the fock representation of a Gaussian tensor.
+    This implementation works on flattened tensors and reshapes the tensor before returning.
+
+    The vanilla algorithm implements the flattened version of the following recursion which
+    calculates the Fock amplitude at index :math:`k` using all available pivots at index :math:`k - 1_i`
+    for all :math:`i`, and their neighbours at indices :math:`k - 1_i - 1_j`:
+
+    .. math::
+
+         G_{k} = \frac{1}{N}\sum_i\frac{1}{\sqrt{k_i}} \left[b_{i} G_{k-1_i} + \sum_j A_{ij} \sqrt{k_j - \delta_{ij}} G_{k-1_i-1_j} \right]
+
+    where :math:`N` is the number of valid pivots for the :math:`k`-th lattice point, :math:`1_i` is
+    the vector of zeros with a 1 at index :math:`i`, and :math:`\delta_{ij}` is the Kronecker delta.
+    In the implementation the indices are flattened into a single integer index, which makes the
+    computations of the pivots more efficient.
+
+    see https://quantum-journal.org/papers/q-2020-11-30-366/ and
+    https://arxiv.org/abs/2209.06069 for more details.
 
     Args:
-        shape: shape of the output tensor excluding the batch dimension
-        A: A matrix of the Fock-Bargmann representation
-        b: B vector of the Fock-Bargmann representation (eventually batched with batch on the first dimension)
-        c: vacuum amplitude
+        shape (tuple[int, ...]): shape of the output tensor
+        A (np.ndarray): A matrix of the Bargmann representation
+        b (np.ndarray): b vector of the Bargmann representation
+        c (complex): vacuum amplitude
 
     Returns:
-        np.ndarray: Fock representation of the Gaussian tensor with shape ``(batch,) + shape``
+        np.ndarray: Fock representation of the Gaussian tensor with shape ``shape``
     """
-    A = np.array(A)
-    b = np.array(b)
-    c = np.array(c)
-    if b.ndim == 1:
-        b = np.atleast_2d(b)
-        return _vanilla_average_batch(shape, A, b, c)[..., 0]
-    elif b.ndim == 2:
-        return np.moveaxis(_vanilla_average_batch(shape, A, b, c), -1, 0)
-    else:
-        raise ValueError(f"Invalid shape for b: {b.shape}. It should be 1D or 2D.")
+    D = b.shape[0]
```

Comment:
Why does this not mirror the same way that vanilla starts?

Reply:
No particular reason, I wrote them on different days 😄